ADR/0033-provisioning-test-resources.md

```diff
@@ -0,0 +1,279 @@
+aid of dependabot) and deployed to a single Konflux-managed namespace. `ClusterClaims` must reside
+in the same namespace as the `ClusterPools` but that's also where the cloud secrets are stored.
+Furthermore, the user cannot be allowed to create claims in their tenant namespace since it will
+only generate errors from Hive due to missing cluster pools. We could grant each tenant's pipeline
+service account the permission to create claims in the protected namespace but that opens up an
+attack surface where the pool of clusters could be starved by scripted creation of claims.
+Alternatively, something else could create the claims in the protected namespace on behalf of the
+user. A DRA provider would have been this actor.
+
+### Declaring Ephemeral Resources
+
+A `provision.konflux.dev/cluster-claims` annotation on integration `Pipelines`/`PipelineRuns` will
+be used to signal the intent to create claims for ephemeral OpenShift clusters. The value of this
+annotation is a reference to a `ConfigMap` in the same namespace. The `ConfigMap` keys will be used
+to dynamically name certain resources. The values are `ClusterClaims` to be applied alongside the
+managed `ClusterPools`.
+
+A new controller will apply the `ClusterClaims` to the namespace containing the pools, not the
+tenant's namespace. The claims must reference the predefined `ClusterPools` or the user
+should expect to encounter errors (via conditions on the `PipelineRun`). The name of the
+`ClusterClaim` will be replaced using the pattern `<PipelineRunName>-<ConfigMapKey>` and
+additional labels/annotations may be used to keep track of the origin tenant details. The
+`lifetime` of claims may be defined by the user but an upper bound can be enforced at this
+layer (e.g. via a controller config or dynamically based on the timeout of the `PipelineRun`).
+Hive will take care of deleting claims and their associated clusters when the lifetime has been
+exceeded. Additionally, this controller may delete `ClusterClaims` as `PipelineRuns` reach
+completion or get pruned/deleted.
+
+If desired, the controller can be enhanced to support deploying additional resources at the start
+of a `PipelineRun`. Take these potential annotations, for example:
+* `provision.konflux.dev/resource-claims` might reference a `ConfigMap` of DRA `ResourceClaims`
+* `provision.konflux.dev/capi-clusters` might reference a `ConfigMap` of Cluster API `Clusters`
+
+It may be possible to replace this annotation approach if Tekton ever adds support for declaring DRA
+`ResourceClaimTemplates` on a `Pipeline`. However, this isn't on any roadmap at the time of this
+writing.
+
+### Sharing Ephemeral Resources Across `TaskRuns`/`Pods`
+
+The controller will also watch the status of `ClusterClaims` as they are allocated. Upon success,
+the secret for the `ClusterDeployment` (containing a kubeconfig) will be copied to the tenant
+namespace using the same name as the `ClusterClaim`. Using this pattern allows `Pipeline` authors to
```

Comment:
Does this mean that only a single pipeline can be running at a time? Otherwise they would both get the same cluster. 

The multi-arch controller uses a similar pattern but uses the task/pipeline name in the secret, so every secret name is unique.

Reply:
No, there can be many claims for a given PipelineRun. It's mentioned earlier that the claims would include the PipelineRun name (`<PipelineRunName>-<ConfigMapKey>`).